### PR TITLE
Simplify Dune_engine.Build_context

### PR DESCRIPTION
**Pull request by ghost**
`stdlib_dir` and `default_ocamlpath` don't really belong there.

### CODE REVIEW

**Comment by jeremiedimino** [comment]
I updated this PR so that the call to `Build_system.init` in `main.ml` no longer uses `Context.DB.all`, which should make things easier when we don't have `Context.t` values at hand.

File: src/dune_rules/workspace.ml [hunk 1]
```diff
@@ -324,6 +324,22 @@ module Context = struct
       ; dynamically_linked_foreign_archives = true
       ; instrument_with = Option.value instrument_with ~default:[]
       }
+
+  let build_contexts t =
+    let name = name t in
+    let native =
+      Build_context.create ~name ~host:(host_context t)
+        ~build_dir:(Path.Build.of_string (Context_name.to_string name))
```
Comment (rgrinberg) [inline src/dune_rules/workspace.ml:332]: There's `Context_name.build_dir` to simplify this.

File: src/dune_rules/workspace.ml [hunk 2]
```diff
@@ -324,6 +324,22 @@ module Context = struct
...
+    in
+    native
+    ::
+    List.filter_map (targets t) ~f:(function
+      | Native -> None
+      | Named toolchain ->
+        let name = Context_name.target name ~toolchain in
+        Some
+          (Build_context.create ~name ~host:(Some native.name)
+             ~build_dir:(Path.Build.of_string (Context_name.to_string name))))
```
Comment (rgrinberg) [inline src/dune_rules/workspace.ml:342]: ditto

File: src/dune_rules/workspace.ml (hunk 1 above)
Comment (rgrinberg) [reply src/dune_rules/workspace.ml:332]: Actually, do we need to pass `build_dir` at all?  `create` can figure it out from the name.

Comment (jeremiedimino) [reply src/dune_rules/workspace.ml:332]: indeed, changing that

**Review by rgrinberg — APPROVED** [review]

**Review by snowleopard — APPROVED** [review]
Looks good to me.

**Pull request merged**